### PR TITLE
[feature][processing] add 'Shortest Line between layers' algorithm in Processing

### PR DESCRIPTION
**Pull request by ghtmtt**
## Description

The algorithm creates a line layer as the result of the shortest line calculation between the source and the destination layer. The shortest line is calculated using the `shortestLine` method, so the distances are purely Cartesian.

The user can specify the numbers of neighbor to calculate the lines from and the optional parameter of the maximum distance can be used to consider only the features of the destination layer within this distance.

image

### CODE REVIEW

**Comment by gioman** [comment]
@ghtmtt nice! what about polygons? is the distance computed to/from boundary? centroid? is there an option to choose? how is speed with large datasets?

**Comment by ghtmtt** [comment]
> @ghtmtt nice! what about polygons? is the distance computed to/from boundary? centroid? is there an option to choose? how is speed with large datasets?

with polygons yes, from/to the boundary and not the centroid. No option provided so far.

image

for the performance I tested with 1000 points against other 1000 points (1 millions lines are created) and the runtime is 1'27''. I've optimized it with the using of the spatial index

**Comment by gioman** [comment]
@ghtmtt nice. For some uses (i.e. landscape analysis/ecology) sometimes is perfectly fine using centroids, and it makes the analysis also much faster. An option would be nice.

**Comment by alexbruy** [comment]
May I suggest to implement it as native C++ algorithm? We are trying to move most of the Processing Python code to C++

File: python/plugins/processing/algs/qgis/ShortestDistance.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
+# -*- coding: utf-8 -*-
+
+"""
+***************************************************************************
+    Gridify.py
```
Comment (nicogodet) [inline python/plugins/processing/algs/qgis/ShortestDistance.py:5]:
```suggestion
    ShortestDistance.py
```

File: python/plugins/processing/algs/qgis/ShortestDistance.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+        if sink is None:
+            raise QgsProcessingException(self.invalidSinkError(parameters, self.OUTPUT))
+
+        request = QgsFeatureRequest()
+        request.setDestinationCrs(source.sourceCrs(), context.transformContext())
+
+        index = QgsSpatialIndex(QgsSpatialIndex.FlagStoreFeatureGeometries)
+        for i in destination.getFeatures(request):
+            index.addFeature(i, QgsFeatureSink.FastInsert)
```
Comment (nicogodet) [inline python/plugins/processing/algs/qgis/ShortestDistance.py:171]:
```suggestion
        index = QgsSpatialIndex(destination.getFeatures(request), flags=QgsSpatialIndex.FlagStoreFeatureGeometries)
```

Doc : 

> This is much faster approach than creating an empty index and then inserting features one by one.

File: python/plugins/processing/algs/qgis/ShortestDistance.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+        total = 100.0 / source.featureCount() if source.featureCount() else 0
+
+        da = QgsDistanceArea()
+        da.setSourceCrs(source.sourceCrs(), context.transformContext())
+        da.setEllipsoid(context.ellipsoid())
+
+        total = 100.0 / source.featureCount() if source.featureCount() else 0
```
Comment (nicogodet) [inline python/plugins/processing/algs/qgis/ShortestDistance.py:179]:
```suggestion
```

Duplicated line

**Comment by ghtmtt** [comment]
> May I suggest to implement it as native C++ algorithm? We are trying to move most of the Processing Python code to C++

@alexbruy I'm sorry but I'm not a C++ developer, so I thought that rather to have a plugin just for this feature, I could added it as a python core script

File: python/plugins/processing/algs/qgis/ShortestDistance.py (hunk 2 above)
Comment (ghtmtt) [reply python/plugins/processing/algs/qgis/ShortestDistance.py:171]: didn't know that, thanks!

**Review by ghtmtt — COMMENTED** [review]
thanks @nicogodet for your review!

**Comment by roya0045** [comment]
Why not name it 'Shortest line'? Adding the 'between layers' made me think that the layer extents were used, instead of being per-feature.

**Comment by nyalldawson** [comment]
I'm with @alexbruy here -- I think the time has come to only accept C++ algorithms in master (or algorithms which absolutely cannot depend on python libraries). We can use this work as a prototype for a C++ version, and in the meantime it could be made available as either or plugin or custom script algorithm.

**Comment by root676** [comment]
Thats a great algorithm - @ghtmtt I would volunteer porting it to C++ in the near future.

**Comment by ghtmtt** [comment]
@nyalldawson @alexbruy I perfectly understand that C++ algorithms have better performances compared to python ones. My personal opinion is that it is more difficult to have C++ contributions than Python ones. 

I strongly agree that having algorithms that depends on external libraries should be avoided and shipped as plugins.

Also sharing scripts, even if we have the chance to use the `Resource Sharing` plugin is not *that easy*: the *recent* addition of https://plugins.qgis.org/styles/ demonstrates it. 

In QGIS Processing we had the `Add script from online collection` (or something similar) that was easy to use for users. But sorry that's maybe off topic.

@root676 thanks for volunteering :) let's see how we are going to proceed :)

**Comment by ghtmtt** [comment]
up to you guys to close or merge this pull request

**Comment by gioman** [comment]
> up to you guys to close or merge this pull request

it is a valuable tool, I would not like to see this closed.

**Comment by ghtmtt** [comment]
today is also feature freeze. If we want to have it in the next release the we should decide shortly

**Pull request closed without merging**